### DISCUSSION TITLE
What is the service for? And how to open files from command line into existing gui?

### DISCUSSION THREAD

**Discussion by patmaddox** [d12frosted/homebrew-emacs-plus#461] [Q&A]
From what I can tell, the service runs an emacs daemon that we can connect to using `emacsclient`, correct? But I think `emacsclient` is restricted to the terminal version, and there's no way to use the mac GUI? (#398 proposes a change for the app).

Now for a more practical question: I have launched emacs GUI from Terminal (because if don't do it from Terminal, I don't have a useful path... seems to be a pretty common issue). Now I want to edit another file in my existing GUI – is there a terminal command that I can run that will open the file in the existing GUI? As of now I copy the file path from terminal, switch to emacs, and open the file from within emacs... I'd like to be able to do `emacs file` (or whatever the command is) and have it load in the GUI. Can I do that?

**Comment by d12frosted** [comment]
Disclaimer: as a person who doesn't use Emacs server I find it difficult to reason about its capabilities and use cases.

From quick testing, I can say that `emacsclient` supports GUI as well. There is an option `-nw` to run in non-GUI mode. To test this, I started `Emacs.app` normally, run `M-x server-start` and then connected to the server from terminal using `emacsclient -s SOCKET some-file`. It opened `some-file` in my GUI Emacs that I used to start the server. Had to put `SOCKET` manually, because I never configured server thing to work manually. I took the value of `server-socket-dir` to find path to the socket file.

> because if don't do it from Terminal, I don't have a useful path... seems to be a pretty common issue

Yes, this is pretty annoying. But this is how `.app` works on macOS. Though for Emacs there are several solutions. A general one is [exec-path-from-shell](https://github.com/purcell/exec-path-from-shell) which updates `PATH` in the `Emacs.app` process. And also 
`emacs-plus@29` now injects your `PATH` into `Info.plist` file of `Emacs.app`, meaning that it should work in most of the cases even without `exec-path-from-shell` (see https://github.com/d12frosted/homebrew-emacs-plus/pull/453).

> Now I want to edit another file in my existing GUI – is there a terminal command that I can run that will open the file in the existing GUI? [...] Can I do that?

`emacsclient` itself :) You just need to configure socket location, so you don't have to manually provide it every time.

P.S. Making server experience better in `emacs-plus` is on my radar.

**Reply by patmaddox** [reply]
> > Now I want to edit another file in my existing GUI – is there a terminal command that I can run that will open the file in the existing GUI? [...] Can I do that?
> 
> `emacsclient` itself :) You just need to configure socket location, so you don't have to manually provide it every time.

Ah I wasn't clear – I thought I would be able to use `brew services start emacs-plus`, and then run `emacsclient myfile` to launch a _GUI_ version of Emacs on that file.

You're right that if I start GUI emacs, and then `m-x emacs-server`, and then `emacsclient myfile` that it will reuse that GUI.

I had assumed that the brew service would produce an "always on" emacs that _also_ supported GUI. It seems that that's not the case, and that the service produces a terminal-only version.

**Comment by mjgiffin** [comment]
Trying to weigh the pros and cons of using the service vs launchd with the included .plist file:
- service will start emacs daemon at boot
- launchd will start emacs daemon at logon
Anything else to consider?

To use emacsclient with the GUI, use emacsclient -c. You can also make an Automator application and put it in /Applications.

**Reply by patmaddox** [reply]
> To use emacsclient with the GUI, use emacsclient -c.

Right you are!

I usually only want one emacs frame. So I'd need to do some sort of check to see if there's a GUI frame, in which case I can reuse it. If not, I should `-c` to make a new one. If there's no GUI frame, then `emacsclient` will make (or connect to?) a terminal frame.
